R/particle_filter_data.R

```diff
@@ -117,37 +119,51 @@ particle_filter_data <- function(data, time, rate, initial_time = NULL,
     model_time_end <- time_split[[1]]
   }
 
+  ## This is only required for discrete time models really
   assert_integer(model_time_end, name = sprintf("data$%s", time))
-  if (!is_continuous && !all(diff(model_time_end) == 1)) {
-    ## It's possible that we can make this work ok for irregular time
-    ## units, but we make this assumption below when working out the
-    ## start and end step (i.e., that we assume that the data
-    stop("Expected each time difference to be one unit")
+
+  if (is.null(initial_time)) {
+    if (is_continuous) {
+      stop("'initial_time' must be given for continuous models")
+    } else {
+      initial_time <- model_time_end[[1L]] - 1
+      fmt <- paste("'initial_time' should be provided. I'm assuming '%d'",
+                   "which is one time unit before the first time in your",
+                   "data (%d), but this might not be appropriate. This",
+                   "will become an error in a future version of mcstate")
+      warning(sprintf(fmt, initial_time, model_time_end[[1L]]),
+              immediate. = TRUE)
+
+      ## This is only an issue while we allow not providing an initial
+      ## time.
+      if (model_time_end[[1L]] < 1) {
+        stop(sprintf("The first time must be at least 1 (but was given %d)",
+                     model_time_end[[1L]]))
+      }
+    }
   }
 
+  ## I am not entirely sure why we require two time windows and not
+  ## one - it's possible this is a hangover an earlier version where
+  ## the first line was the start time?
   if (length(model_time_end) < 2) {
     stop("Expected at least two time windows")
   }
 
-  ## NOTE: test is against 1 because we'll start at 1 - 1 = 0
-  if (!is_continuous && model_time_end[[1L]] < 1) {
-    stop(sprintf("The first time must be at least 1 (but was given %d)",
-                 model_time_end[[1L]]))
+  if (any(model_time_end < 0)) {
+    ## This condition is actually only required for discrete time
+    ## models; for continuous time models this would be fine.
+    stop("All times must be non-negative")
   }
 
-  if (is.null(initial_time)) {
-    if (is_continuous) {
-      stop("'initial_time' must be given for continuous models")
-    }
-    initial_time <- model_time_end[[1L]] - 1
-  } else {
-    initial_time <- assert_integer(initial_time)
-    if (initial_time < 0) {
-      stop("'initial_time' must be non-negative")
-    }
-    if (initial_time > model_time_end[[1L]] - 1) {
-      stop(sprintf("'initial_time' must be <= %d", model_time_end[[1L]] - 1))
-    }
+  initial_time <- assert_integer(initial_time)
+  if (initial_time < 0) {
+    ## This condition is actually only required for discrete time
+    ## models; for continuous time models this would be fine.
+    stop("'initial_time' must be non-negative")
+  }
+  if (initial_time > model_time_end[[1L]]) {
```

Comment:
why was this check previously against `model_time_end[[1L] - 1`?

Reply:
I'm think this was that we used to require at least one time interval to have passed before the first data period, because we used to have the compare function take the current state and the previous state (so that cumulative variables could be converted into differences between reporting periods). But not entirely sure really, and it's certainly a complication that no longer seems required